components/esp_now/esp_now_api.c

```diff
@@ -0,0 +1,300 @@
+// Copyright 2022 PWrInSpace, Kuba
+
+#include "esp_now_api.h"
+
+#include <memory.h>
+
+#include "esp_check.h"
+#include "esp_event.h"
+#include "esp_log.h"
+#include "esp_mac.h"
+#include "esp_netif.h"
+#include "esp_now.h"
+#include "esp_wifi.h"
+#include "nvs_flash.h"
+
+#define TAG "ENA"
+
+
+// how sending works
+// ENA_send
+// -> packet is transmiting -> the packet is added to queue,
+// -> packet is not transmiting -> acquire transmiting flag, add data to transmit param, send data
+// On_receive
+// -> packet transmit succes -> check data on queue (*)
+//                              -> more data -> send_data
+//                              -> no more data -> relese transmit flag
+// -> packet transmit fail -> check retry
+//                              -> more retry -> send_data again
+//                              -> no more retry ->check data on queue (*)
+
+static struct {
+    ENA_device_t **devices;
+    size_t number_of_devices;
+    ENA_on_error error_fnc;
+
+    ENA_transmit_param_t last_packet;
+    // indicate if packet is transmiting, it also block last packet overwrite
+    uint32_t is_packet_transmiting;
+
+    TaskHandle_t api_task;           // task for handling callbacks
+    QueueHandle_t transmit_queue;    // queue for transmiting data
+    QueueHandle_t send_cb_queue;     // queue for handling send callback
+    QueueHandle_t receive_cb_queue;  // queue for handling receive callback
+} gb = {.devices = NULL,
+        .number_of_devices = 0,
+        .error_fnc = NULL,
+        .last_packet = {0},
+        .is_packet_transmiting = ENA_NOT_SENDING,
+        .api_task = NULL,
+        .transmit_queue = NULL,
+        .send_cb_queue = NULL,
+        .receive_cb_queue = NULL};
+
+static void now_receive_cb(const uint8_t *mac, const uint8_t *data, int data_len) {
+  ESP_LOGI(TAG, "Packet received, RSSI, MAC: " MACSTR, MAC2STR(mac));
+  if (mac == NULL || data == NULL || data_len == 0) {
+    ESP_LOGE(TAG, "Argument error :C");
+    gb.error_fnc(ENA_REC);
+    return;
+  }
+
+  ENA_receive_cb_t rec_cb_data;
+  memcpy(&rec_cb_data.src_mac, mac, MAC_ADDRESS_SIZE);
+  memcpy(&rec_cb_data.buffer, data, data_len);
+  rec_cb_data.len = data_len;
+
+  if (xQueueSend(gb.receive_cb_queue, &rec_cb_data, 0) != pdTRUE) {
+    ESP_LOGE(TAG, "Unavle to add receive callback to queue");
+    gb.error_fnc(ENA_REC_CB_QUEUE);
+  }
+}
+
+static void now_send_cb(const uint8_t *mac_addres, esp_now_send_status_t status) {
+  ENA_send_cb_t info;
+
+  info.status = status;
+  memcpy(&info.mac, mac_addres, MAC_ADDRESS_SIZE);
+  if (xQueueSend(gb.send_cb_queue, &info, 0) != pdTRUE) {
+    ESP_LOGE(TAG, "Send callback queeu error");
+    gb.error_fnc(ENA_SEND_CB_QUEUE);
+  }
+}
+
+static bool address_compare(const uint8_t *addr1, const uint8_t *addr2) {
```

Comment:
mozna tu memcmp po prostu zastosowac

Reply:
nie wiem co jest w środku, tak to wiem że przy pierwszej różnicy skończy iterować